File: src/ru/nsu/ccfit/zuev/osu/beatmap/parser/sections/BeatmapMetadataParser.java
package ru.nsu.ccfit.zuev.osu.beatmap.parser.sections;

import ru.nsu.ccfit.zuev.osu.beatmap.BeatmapData;

/**
 * A parser for parsing a beatmap's metadata section.
 */
public class BeatmapMetadataParser extends BeatmapKeyValueSectionParser {
    @Override
    public void parse(BeatmapData data, String line) {
        String[] p = splitProperty(line);

        switch (p[0]) {
            case "Title" -> data.metadata.title = p[1];
            case "TitleUnicode" -> data.metadata.titleUnicode = p[1];
            case "Artist" -> data.metadata.artist = p[1];
            case "ArtistUnicode" -> data.metadata.artistUnicode = p[1];
            case "Creator" -> data.metadata.creator = p[1];
            case "Version" -> data.metadata.version = p[1];
            case "Source" -> data.metadata.source = p[1];
            case "Tags" -> data.metadata.tags = p[1];
            case "BeatmapID" -> data.metadata.beatmapID = parseInt(p[1]);
            case "BeatmapSetID" -> data.metadata.beatmapSetID = parseInt(p[1]);
        }
    }
}

File: src/com/rian/osu/beatmap/parser/sections/BeatmapMetadataParser.kt
package com.rian.osu.beatmap.parser.sections

import com.rian.osu.beatmap.Beatmap
import kotlinx.coroutines.CoroutineScope

/**
 * A parser for parsing a beatmap's metadata section.
 */
object BeatmapMetadataParser : BeatmapKeyValueSectionParser() {
    override fun parse(beatmap: Beatmap, line: String, scope: CoroutineScope?) = splitProperty(line, scope)?.let {
        when (it.first) {
            "Title" -> beatmap.metadata.title = it.second
            "TitleUnicode" -> beatmap.metadata.titleUnicode = it.second
            "Artist" -> beatmap.metadata.artist = it.second
            "ArtistUnicode" -> beatmap.metadata.artistUnicode = it.second
            "Creator" -> beatmap.metadata.creator = it.second
            "Version" -> beatmap.metadata.version = it.second
            "Source" -> beatmap.metadata.source = it.second
            "Tags" -> beatmap.metadata.tags = it.second
            "BeatmapID" -> beatmap.metadata.beatmapId = parseInt(it.second)
            "BeatmapSetID" -> beatmap.metadata.beatmapSetId = parseInt(it.second)
        }
    } ?: throw UnsupportedOperationException("Malformed metadata property: $line")
}
